library/ssl_srv.c

```diff
@@ -2765,6 +2765,53 @@ static int ssl_write_hello_verify_request( mbedtls_ssl_context *ssl )
 }
 #endif /* MBEDTLS_SSL_DTLS_HELLO_VERIFY */
 
+static void ssl_check_id_based_session_resumption( mbedtls_ssl_context *ssl )
+{
+    int ret;
+    mbedtls_ssl_session session_tmp = { 0 };
+    mbedtls_ssl_session * const session = ssl->session_negotiate;
+
+    /* Resume is 0  by default, see ssl_handshake_init().
+     * It may be already set to 1 by ssl_parse_session_ticket_ext(). */
+    if( ssl->handshake->resume == 1 )
+        return;
+    if( ssl->session_negotiate->id_len == 0 )
+        return;
+    if( ssl->conf->f_get_cache == NULL )
+        return;
+#if defined(MBEDTLS_SSL_RENEGOTIATION)
+    if( ssl->renego_status != MBEDTLS_SSL_INITIAL_HANDSHAKE )
+        return;
+#endif
+
+    session_tmp.id_len = session->id_len;
+    memcpy( session_tmp.id, session->id, session->id_len );
+
+    ret = ssl->conf->f_get_cache( ssl->conf->p_cache,
+                                  &session_tmp );
+    if( ret != 0 )
+        goto exit;
+
+    if( session->ciphersuite != session_tmp.ciphersuite ||
+        session->compression != session_tmp.compression )
+    {
+        /* Mismatch between cached and negotiated session */
+        goto exit;
+    }
+
+    /* Move semantics */
+    /* Zeroization of session_tmp happens at the end of the function. */
+    mbedtls_ssl_session_free( session );
+    *session = session_tmp;
```

Comment:
Hm, I thought that "move semantics" would make this clear.

Reply:
I didn't get it from "move semantics" either. For me seeing a zeroisation of `session_tmp` immediately on the next line would help to read it. Also, if it is a simple `memset()` it will be optimised away if not necessary and would minimise the window of opportunity for a double free.

Reply:
@yanesca @mpg Thanks! I agree, we shouldn't try to be clever and break up the usual move semantics steps of (a) freeing target, (b) shallow-copying, (c) removing any references from source (and in fact, df56402 left the code vulnerable to a double-free). I've added a `memset()` on `session_tmp` after the copy operation. The call to `mbedtls_ssl_session_free()` at the end remains, both to avoid memory leaks upon failure to resume the session, but also to reliably zeroize the temporary structure. As @yanesca said, a clever compiler might even realize that the `memset()` isn't necessary.